### PR TITLE
ARROW-12021: [C++][Gandiva] Implement to_char() function on Gandiva

### PR DESCRIPTION
**Pull request by rodrigojdebem**
Add function to transform numeric values to its equivalent in characters

### CODE REVIEW

**Comment by emkornfield** [comment]
Is this a duplicate of: https://github.com/apache/arrow/pull/8158/files?

**Comment by rodrigojdebem** [comment]
> Is this a duplicate of: https://github.com/apache/arrow/pull/8158/files?

@emkornfield the idea of `to_char` is to allow a pattern to be passed as well so the number is converted to a string and also formatted.

**Comment by emkornfield** [comment]
OK, this PR lacks any documentation to that affect, especially around pattern.  Is this meant to emulate sql server or Oracle's to_char?

**Comment by emkornfield** [comment]
The JIRA for this PR looks like it might map to the wrong issue?

**Comment by rodrigojdebem** [comment]
> OK, this PR lacks any documentation to that affect, especially around pattern. Is this meant to emulate sql server or Oracle's to_char?

@emkornfield I'll be implementing the Oracle `to_char`. Thanks for the heads up about the documentation and JIRA issue, I'll be fixing those as well!

**Comment by rodrigojdebem** [comment]
@emkornfield hello! Do you know the reason for the checks being skipped? Is it because the PR is old?

**Comment by projjal** [comment]
> @emkornfield hello! Do you know the reason for the checks being skipped? Is it because the PR is old?

Maybe the WIP tag?

File: cpp/src/gandiva/function_registry_string.cc [hunk 1]
```diff
@@ -114,6 +114,22 @@ std::vector<NativeFunction> GetStringFunctionRegistry() {
                      utf8(), kResultNullIfNull, "substr_utf8_int64_int64",
                      NativeFunction::kNeedsContext),
 
+      NativeFunction("to_char", {}, DataTypeVector{int32(), int64()}, utf8(),
```
Comment (projjal) [inline cpp/src/gandiva/function_registry_string.cc:117]: The signature looks incorrect. Aren't you taking pattern (string) as second argument.
It should be {int32(), utf8()}

File: cpp/src/gandiva/gdv_function_stubs.cc [hunk 2]
```diff
@@ -176,6 +180,49 @@ char* gdv_fn_dec_to_string(int64_t context, int64_t x_high, uint64_t x_low,
     return val;                                                                      \
   }
 
+#define GDV_FN_TO_CHAR(IN_TYPE, ARROW_TYPE)                                         \
+  GANDIVA_EXPORT                                                                         \
+  const char* gdv_fn_to_char_##IN_TYPE##_int64(int64_t context, gdv_##IN_TYPE value, \
+                                                   int64_t len, const char* pattern, int32_t * out_len) {     \
```
Comment (projjal) [inline cpp/src/gandiva/gdv_function_stubs.cc:186]: This signature is also different.

File: cpp/src/gandiva/gdv_function_stubs.cc [hunk 3]
```diff
@@ -337,6 +384,42 @@ void ExportedStubFunctions::AddMappings(Engine* engine) const {
 
   engine->AddGlobalMappingForFunc("gdv_fn_castFLOAT8_utf8", types->double_type(), args,
                                   reinterpret_cast<void*>(gdv_fn_castFLOAT8_utf8));
+
+    // gdv_fn_to_char_int32_int64
+    args = {types->i64_type(),       // int64_t execution_context
+            types->i32_type(),       // int32_t value
+            types->i64_type(),       // int64_t len
+            types->i32_ptr_type()};  // int32_t* out_len
+    engine->AddGlobalMappingForFunc(
```
Comment (projjal) [inline cpp/src/gandiva/gdv_function_stubs.cc:393]: This arguments here are also different to your actual function.

File: cpp/src/gandiva/gdv_function_stubs_test.cc [hunk 4]
```diff
@@ -160,4 +160,118 @@ TEST(TestGdvFnStubs, TestCastFloat8) {
   ctx.Reset();
 }
 
+    TEST(TestGdvFnStubs, TestToCharFromInt32) {
+        gandiva::ExecutionContext ctx;
+        uint64_t ctx_ptr = reinterpret_cast<int64_t>(&ctx);
+        int32_t out_len = 0;
+
+        const char* out_str = gdv_fn_to_char_int32_int64(ctx_ptr, -46, 100, "%1", &out_len);
```
Comment (emkornfield) [inline cpp/src/gandiva/gdv_function_stubs_test.cc:168]: Can you check some other examples for format.  I'm not sure the library that is being used uses the same specification for format strings a SQL to_char would.

File: cpp/src/gandiva/gdv_function_stubs.cc [hunk 5]
```diff
@@ -302,40 +302,41 @@ char* gdv_fn_dec_to_string(int64_t context, int64_t x_high, uint64_t x_low,
     return val;                                                                      \
   }
 
-#define GDV_FN_TO_CHAR(IN_TYPE, ARROW_TYPE)                                         \
-  GANDIVA_EXPORT                                                                         \
+#define GDV_FN_TO_CHAR(IN_TYPE, ARROW_TYPE)                                          \
```
Comment (emkornfield) [inline cpp/src/gandiva/gdv_function_stubs.cc:305]: there should be comments on these macros to describe what the function does.

File: cpp/src/gandiva/gdv_function_stubs_test.cc (hunk 4 above)
Comment (rodrigojdebem) [reply cpp/src/gandiva/gdv_function_stubs_test.cc]: The implementation is aiming to use the Java formatting.

Comment (emkornfield) [reply cpp/src/gandiva/gdv_function_stubs_test.cc]: I thought this was mean to mimic oracle's to_char?  if Java formatting is the target, maybe include more examples from that?  I'm not sure the boost formatting library has the same formatting options with the same characters.

File: cpp/src/gandiva/gdv_function_stubs.cc [hunk 6]
```diff
@@ -300,6 +302,50 @@ char* gdv_fn_dec_to_string(int64_t context, int64_t x_high, uint64_t x_low,
     return val;                                                                      \
   }
 
+#define GDV_FN_TO_CHAR(IN_TYPE, ARROW_TYPE)                                         \
+  /* Converts expression to a string using the format specified in format */                                                                                  \
+  GANDIVA_EXPORT                                                                         \
+  const char* gdv_fn_to_char_##IN_TYPE##_int64(int64_t context, gdv_##IN_TYPE value, \
+                                                   int64_t len, const char* pattern, int32_t * out_len) {     \
```
Comment (projjal) [inline cpp/src/gandiva/gdv_function_stubs.cc:309]: You are not using pattern at all in your code. Doesn't look like it should be working.

File: cpp/src/gandiva/gdv_function_stubs.cc [hunk 7]
```diff
@@ -300,6 +302,50 @@ char* gdv_fn_dec_to_string(int64_t context, int64_t x_high, uint64_t x_low,
     return val;                                                                      \
   }
 
+#define GDV_FN_TO_CHAR(IN_TYPE, ARROW_TYPE)                                         \
+  /* Converts expression to a string using the format specified in format */                                                                                  \
```
Comment (projjal) [inline cpp/src/gandiva/gdv_function_stubs.cc:306]: Can you write a more detailed comment explaining the format?

File: cpp/src/gandiva/gdv_function_stubs_test.cc [hunk 8]
```diff
@@ -290,4 +290,121 @@ TEST(TestGdvFnStubs, TestCastVARCHARFromDouble) {
...
+  ctx.Reset();
+}
+
+TEST(TestGdvFnStubs, TestToCharFromInt64) {
+  gandiva::ExecutionContext ctx;
+  uint64_t ctx_ptr = reinterpret_cast<int64_t>(&ctx);
+  int32_t out_len = 0;
+
+  const char* out_str =
+      gdv_fn_to_char_int64_int64(ctx_ptr, 9223372036854775807LL, 100, "%1", &out_len);
```
Comment (projjal) [inline cpp/src/gandiva/gdv_function_stubs_test.cc:335]: what does %1 do?

File: cpp/src/gandiva/gdv_function_stubs.cc (hunk 6 above)
Comment (rodrigojdebem) [reply cpp/src/gandiva/gdv_function_stubs.cc:309]: You're right. I lost something among my commits and merges. Looking into it now.

File: cpp/src/gandiva/gdv_function_stubs.cc [hunk 9]
```diff
@@ -307,6 +310,21 @@ CAST_NUMERIC_FROM_STRING(double, arrow::DoubleType, FLOAT8)
...
+    gandiva::ToCharHolder* holder = reinterpret_cast<gandiva::ToCharHolder*>(context); \
+    return holder->Format<gdv_##IN_TYPE>(data);                                        \
+  }
+
+GDV_FN_TO_CHAR(int32, arrow::Int32Type, INT)
+GDV_FN_TO_CHAR(int64, arrow::Int64Type, BIGINT)
+GDV_FN_TO_CHAR(float32, arrow::FloatType, FLOAT4)
+GDV_FN_TO_CHAR(float64, arrow::DoubleType, FLOAT8)
+
+#undef CAST_NUMERIC_FROM_STRING
```
Comment (vvellanki) [inline cpp/src/gandiva/gdv_function_stubs.cc:326]: Shouldn't this be #undef GDV_FN_TO_CHAR?

File: cpp/src/gandiva/to_char_holder.h [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+  // Sets the format's metadata, such as the maximum number of decimal digits and if
+  // the patterns contains a dollar sign.
+  int32_t Setup() {
+    int32_t ret = 0;
+    bool is_decimal_part = false;
+    has_dollar_sign_ = false;
+    for (size_t i = 0; i < pattern_size_; ++i) {
+      if (pattern_[i] == '$') {
+        has_dollar_sign_ = true;
```
Comment (vvellanki) [inline cpp/src/gandiva/to_char_holder.h:58]: Add continue here

**Pull request closed without merging**